src/System.Xml.XDocument/src/System/Xml/Linq/OrderedNodeEnumerable.cs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System.Xml.Linq
+{
+    internal sealed class OrderedNodeEnumerable<TElement> : IEnumerable<TElement> where TElement : XNode
+    {
+        private readonly IEnumerable<TElement> _source;
+
+        internal OrderedNodeEnumerable(IEnumerable<TElement> source)
+        {
+            if (source == null) throw new ArgumentNullException("source");
+            _source = source;
+        }
+        
+        public IEnumerator<TElement> GetEnumerator()
+        {
+            TElement[] items;
+            int count;
+            ICollection<TElement> collection = _source as ICollection<TElement>;
+            if (collection != null)
+            {
+                count = collection.Count;
+                if (count == 0) yield break;
+                items = new TElement[count];
+                collection.CopyTo(items, 0);
+            }
+            else
+            {
+                using (IEnumerator<TElement> e = _source.GetEnumerator())
+                {
+                    if (!e.MoveNext()) yield break;
+                    items = new TElement[4];
+                    items[0] = e.Current;
+                    count = 1;
+                    
+                    while (e.MoveNext())
+                    {
+                        if (items.Length == count) Array.Resize(ref items, checked(count * 2));
+                        items[count] = e.Current;
+                        ++count;
+                    }
+                }
+            }
+            QuickSort(items, 0, count - 1, count);
```

Comment:
Is there a reason this can't use `Array.Sort(items, 0, count - 1, XNode.DocumentOrderComparer)`?